lib/internal/modules/esm/loader.js

```diff
@@ -96,6 +100,59 @@ const {
 
 let emittedSpecifierResolutionWarning = false;
 
+function nextHookFactory(chain, meta, validate) {
+  // First, prepare the current
+  const { hookName } = meta;
+  const {
+    fn: hook,
+    url: hookFilePath,
+  } = chain[meta.hookIndex];
+
+  const nextHookName = `next${
+    StringPrototypeToUpperCase(hookName[0]) +
+    StringPrototypeSlice(hookName, 1)
+  }`;
+
+  // When hookIndex is 0, it's reached the default, which does not call next()
+  // so feed it a noop that blows up if called, so the problem is obvious.
+  const generatedHookIndex = meta.hookIndex;
+  let nextNextHook;
+  if (meta.hookIndex > 0) {
+    // Now, prepare the next: decrement the pointer so the next call to the
+    // factory generates the next link in the chain.
+    meta.hookIndex--;
+
+    nextNextHook = nextHookFactory(chain, meta, validate);
+  } else {
+    // eslint-disable-next-line func-name-matching
+    nextNextHook = function chainAdvancedTooFar() {
+      throw new RangeError(
+        `ESM custom loader '${hookName}' advanced beyond the end of the chain; this is a bug in Node.js itself.`
+      );
+    };
+  }
+
+  return ObjectDefineProperty(
+    async (...args) => {
+      // Update only when hook is invoked to avoid fingering the wrong filePath
+      meta.hookErrIdentifier = `${hookFilePath} '${hookName}'`;
+
+      validate(meta.hookErrIdentifier, ...args);
```

Comment:
Can we change the `validate` signature to avoid the array iteration here?
```suggestion
      validate(meta.hookErrIdentifier, args);
```

Reply:
Ah, damn. I did it this way because it looked cleaner, but user-land gah. Yeah will do, and then destructure in the declaration (`validate(hookErrIdentifier, { 0: specifier, 1: context })`.

Reply:
Oh, wait, actually: Does it matter that it can't be avoided for line 145 below:

https://github.com/nodejs/node/blob/9099755fb6c9447b1e3283d63292786cb209a379/lib/internal/modules/esm/loader.js#L145

(`nextHookFactory` can't be aware of what those args are)

Reply:
That was going to be my next suggestion: to remove the spreading there too.

Reply:
I think it can't because it would change the signature of custom loader's `resolve` and `load` hooks: `resolve(specifier, context, nextResolve)` → `resolve([specifier, context], nextResolve)`

Reply:
Both `resolve` and `load` _happen_ to both have 2 args in there (so I _could_ destructure them as `{0: arg1, 1: arg2}`), but there's nothing saying it has to stay that way and I think doesn't make too much sense (more of a foot-gun if I did do that `arg1` `arg2` thing).

Reply:
My suggestion is not to change the hooks signatures, but to push `nextNextHook` to `args` and use `ReflectApply` instead.

Reply:
OH! Brilliant. Done